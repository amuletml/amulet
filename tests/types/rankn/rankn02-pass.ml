type id 'a = Id of 'a

let unId (Id x) = x

class typeable 'a => data 'a begin
  val gfoldl : (forall 'a 'b. data 'a => 'w ('a -> 'b) -> 'a -> 'w 'b)
            -> (forall 'g. 'g -> 'w 'g)
            -> 'a -> 'w 'a

  val gmapT : (forall 'b. data 'b => unit -> 'b -> 'b) -> 'a -> 'a
  let gmapT f x =
    unId (gfoldl (fun (Id c) x -> Id (c (f () x))) Id x)
end

instance data int begin
  let gfoldl _ z x = z x
end
