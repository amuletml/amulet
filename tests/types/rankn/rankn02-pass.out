id : type -> type
Id : Spec{'a : type}. 'a -> id 'a
unId : Infer{'a : type}. id 'a -> 'a
data : type -> constraint
gmapT : Spec{'a : type}. data 'a => (Spec{'b : type}. data 'b => unit -> 'b -> 'b) -> 'a -> 'a
gfoldl : Spec{'a : type}. data 'a => Spec{'w : type -> type}. (Spec{'a : type}. Spec{'b : type}. data 'a => 'w ('a -> 'b) -> 'a -> 'w 'b) -> (Spec{'g : type}. 'g -> 'w 'g) -> 'a -> 'w 'a
